File: tests/memtest/mem_test.f90
! Memory test for the azul Fortran binding. See tests/memtest/README.md.
!
! The harness (scripts/run_memtest.sh) measures peak RSS across a small and a
! large AZ_MEMTEST_N (RSS that scales with N is a LEAK) and fails on any crash.
! This file only exercises the create/consume/DROP paths in a loop and exits 0.
! No event loop (app%run needs a display and hangs headless).
!
! Uses the wrapper layer, like examples/fortran. Build with the shipped
! Makefile.fortran (drop azul.f90 + libazul.so next to this file).

program mem_test
  use azul
  implicit none

  type :: t_model
    integer :: counter
  end type t_model

  integer :: n, i, ios, ln
  type(app_t) :: the_app
  type(app_config_t) :: cfg
  character(len=32) :: nval

  n = 200000
  call get_environment_variable('AZ_MEMTEST_N', nval, ln)
  if (ln > 0) then
    read(nval, *, iostat=ios) i
    if (ios == 0) n = i
  end if

  ! 1. The consume-by-value DROP path: app_create moves the AppConfig bytes
  !    (nested SystemStyle) into libazul; the%delete drops the App once. The
  !    RefAny owns a binding-side copy of t_model, freed by the handle-table
  !    releaser when libazul drops the last clone.
  the_app = app_create(ref_any_create(t_model(5)), app_config_create())
  call the_app%delete()

  ! 2. Leak loop: create/destroy a droppable AppConfig N times.
  !    app_config_delete drops the nested SystemStyle every iteration.
  do i = 1, n
    cfg = app_config_create()
    call cfg%delete()
  end do

  print '(A,I0,A)', 'memtest fortran OK (N=', n, ')'
end program mem_test

File: tests/memtest/mem_test.pas
{ Memory test for the azul Pascal (FPC) binding. See tests/memtest/README.md.

  The harness (scripts/run_memtest.sh) measures peak RSS across a small and a
  large AZ_MEMTEST_N (RSS that scales with N is a LEAK) and fails on any crash.
  This file only exercises the create/consume/DROP paths in a loop and exits 0.
  No event loop (AzApp_run needs a display and hangs headless).

  Build, mirroring examples/pascal:
    fpc -Mobjfpc -Sh -Fl. mem_test.pas && LD_LIBRARY_PATH=. ./mem_test }

program MemTest;

{$mode objfpc}{$H+}
{$PACKRECORDS C}

uses
  ctypes, sysutils,
  Azul;

var
  n, i: Integer;
  model: TObject;
  data: TAzRefAny;
  cfg: TAzAppConfig;
  app: TAzApp;
  envN: string;

begin
  n := 200000;
  envN := GetEnvironmentVariable('AZ_MEMTEST_N');
  if envN <> '' then
    n := StrToIntDef(Trim(envN), 200000);

  { 1. The consume-by-value DROP path: AzApp_create moves the AppConfig bytes
       (nested SystemStyle) into libazul; AzApp_delete drops the App once. }
  model := TObject.Create;
  data := azul_refany_create(model);
  cfg := AzAppConfig_create();
  app := AzApp_create(data, cfg);
  AzApp_delete(@app);

  { 2. Leak loop: create/destroy a droppable AppConfig N times.
       AzAppConfig_delete drops the nested SystemStyle every iteration. }
  for i := 1 to n do
  begin
    cfg := AzAppConfig_create();
    AzAppConfig_delete(@cfg);
  end;

  WriteLn(Format('memtest pascal OK (N=%d)', [n]));
end.
